// exiftool-vendored/dist/DateTime.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.toExifString = exports.dateTimeToExif = exports.isDateOrTime = exports.DayMs = exports.HourMs = exports.MinuteMs = exports.validDateTime = void 0;
const luxon_1 = require("luxon");
const ExifDate_1 = require("./ExifDate");
const ExifDateTime_1 = require("./ExifDateTime");
const ExifTime_1 = require("./ExifTime");
function validDateTime(dt) {
    return dt != null && dt.isValid;
}
exports.validDateTime = validDateTime;
exports.MinuteMs = 60 * 1000;
exports.HourMs = 60 * exports.MinuteMs;
exports.DayMs = 24 * exports.HourMs;
function isDateOrTime(o) {
    return (o instanceof ExifDateTime_1.ExifDateTime ||
        o instanceof ExifDate_1.ExifDate ||
        o instanceof ExifTime_1.ExifTime ||
        o instanceof luxon_1.DateTime);
}
exports.isDateOrTime = isDateOrTime;
function dateTimeToExif(d) {
    return d.toFormat("y:LL:dd HH:mm:ss.u");
}
exports.dateTimeToExif = dateTimeToExif;
function toExifString(d) {
    if (d instanceof luxon_1.DateTime) {
        return dateTimeToExif(d);
    }
    else {
        return d.toExifString();
    }
}
exports.toExifString = toExifString;
